File: backend/app/schemas/metrics.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field

from app.schemas.common import AgentLensSchema
from app.schemas.enums import ScenarioId
from app.schemas.provider import ProviderFeedSummary, ProviderTotalSummary


class OverviewResponse(AgentLensSchema):
    generated_at: datetime
    active_scenario_id: ScenarioId
    agent_count: int = Field(ge=0)
    total_shared_cash_minor: int = Field(ge=0)
    provider_totals: list[ProviderTotalSummary]
    feed_summary: list[ProviderFeedSummary]
    is_synthetic_data: bool = True


class MetricMetadata(AgentLensSchema):
    availability: Literal["AVAILABLE", "UNAVAILABLE"]
    sample_count: int = Field(ge=0)
    measured_at: datetime | None
    version: str | None


class ForecastMetricGroup(MetricMetadata):
    mae_net_outflow_minor: float | None
    rmse_net_outflow_minor: float | None
    smape_percent: float | None
    shortage_detection_lead_time_minutes: float | None
    shortage_lead_time_sample_count: int | None


class AnomalyMetricGroup(MetricMetadata):
    precision: float | None
    recall: float | None
    f1: float | None
    false_positive_rate: float | None
    contextual_false_positive_rate: float | None
    evidence_coverage: float | None


class AIMetricGroup(MetricMetadata):
    completed_count: int | None
    failed_count: int | None
    blocked_count: int | None
    fallback_count: int | None
    average_latency_ms: float | None
    validation_pass_count: int | None
    validation_failure_count: int | None
    source_coverage_rate: float | None
    one_call_evidence_available: bool
    one_call_compliance_rate: float | None


class WorkflowMetricGroup(MetricMetadata):
    case_counts: dict[str, int] | None
    decision_count: int | None
    average_acknowledgement_seconds: float | None
    average_resolution_seconds: float | None
    resolution_rate: float | None
    dismissal_rate: float | None


class MetricsResponse(AgentLensSchema):
    generated_at: datetime
    forecast: ForecastMetricGroup
    anomaly: AnomalyMetricGroup
    ai: AIMetricGroup
    workflow: WorkflowMetricGroup

File: frontend/src/lib/api/metrics.ts
import { metricsSnapshot } from "@/mocks";
import type { MetricMetadata, MetricsSnapshot } from "@/types";
import { mockResponse } from "./mock-client";
import { apiConfig } from "./config";
import { fastApiClient } from "./fastapi-client";
import type { MetricsDto } from "./backend-dto";

const metadata = (value: { availability: "AVAILABLE" | "UNAVAILABLE"; sample_count: number; measured_at: string | null; version: string | null }): MetricMetadata => ({ availability: value.availability, sampleCount: value.sample_count, measuredAt: value.measured_at, version: value.version });
const ratio = (part: number | null, total: number | null): number | null => part !== null && total !== null && total > 0 ? part / total : null;

export async function getMetrics(): Promise<MetricsSnapshot> {
  if (apiConfig.mode === "mock") return mockResponse(metricsSnapshot);
  const value = await fastApiClient.metrics<MetricsDto>();
  return {
    generatedAt: value.generated_at,
    forecast: { ...metadata(value.forecast), maeMinor: value.forecast.mae_net_outflow_minor, rmseMinor: value.forecast.rmse_net_outflow_minor, smapePercent: value.forecast.smape_percent, shortageLeadTimeMinutes: value.forecast.shortage_detection_lead_time_minutes, shortageLeadTimeSampleCount: value.forecast.shortage_lead_time_sample_count, baselineImprovement: null },
    anomaly: { ...metadata(value.anomaly), precision: value.anomaly.precision, recall: value.anomaly.recall, f1Score: value.anomaly.f1, falsePositiveRate: value.anomaly.false_positive_rate, contextualFalsePositiveRate: value.anomaly.contextual_false_positive_rate, evidenceCoverage: value.anomaly.evidence_coverage },
    advisory: { ...metadata(value.ai), structuredOutputValidationRate: ratio(value.ai.validation_pass_count, value.ai.completed_count), sourceReferenceCoverage: value.ai.source_coverage_rate, oneCallEvidenceAvailable: value.ai.one_call_evidence_available, oneCallComplianceRate: value.ai.one_call_compliance_rate, fallbackRate: ratio(value.ai.fallback_count, value.ai.sample_count), averageLatencyMilliseconds: value.ai.average_latency_ms },
    workflow: { ...metadata(value.workflow), averageAcknowledgementMinutes: value.workflow.average_acknowledgement_seconds === null ? null : value.workflow.average_acknowledgement_seconds / 60, averageResolutionMinutes: value.workflow.average_resolution_seconds === null ? null : value.workflow.average_resolution_seconds / 60, decisionCount: value.workflow.decision_count, resolutionRate: value.workflow.resolution_rate, dismissalRate: value.workflow.dismissal_rate, caseCounts: value.workflow.case_counts },
  };
}
